File: typescript/custom-resource-provider/my-custom-resource.ts
<<<<<<< HEAD
import * as cdk from '@aws-cdk/core';
import * as logs from '@aws-cdk/aws-logs';
import * as cr from '@aws-cdk/custom-resources';
import * as lambda from '@aws-cdk/aws-lambda';
=======
import * as cdk from 'aws-cdk-lib';
import * as logs from 'aws-cdk-lib/aws-logs';
import * as cr from 'aws-cdk-lib/custom-resources';
import * as lambda from 'aws-cdk-lib/aws-lambda';
import { Construct } from 'constructs';
>>>>>>> 752df6302e534a7df22a809536c588deaf444c4d
import fs = require('fs');

export interface MyCustomResourceProps {
  /**
   * Message to echo
   */
  Message: string;
}

<<<<<<< HEAD
export class MyCustomResource extends cdk.Construct {
  public readonly response: string;

  constructor(scope: cdk.Construct, id: string, props: MyCustomResourceProps) {
=======
export class MyCustomResource extends Construct {
  public readonly response: string;

  constructor(scope: Construct, id: string, props: MyCustomResourceProps) {
>>>>>>> 752df6302e534a7df22a809536c588deaf444c4d
    super(scope, id);


    const onEvent = new lambda.SingletonFunction(this, 'Singleton', {
      uuid: 'f7d4f730-4ee1-11e8-9c2d-fa7ae01bbebc',
      code: new lambda.InlineCode(fs.readFileSync('custom-resource-handler.py', { encoding: 'utf-8' })),
      handler: 'index.on_event',
      timeout: cdk.Duration.seconds(300),
      runtime: lambda.Runtime.PYTHON_3_6,
    });

    const myProvider = new cr.Provider(this, 'MyProvider', {
      onEventHandler: onEvent,
      // isCompleteHandler: isComplete,        // optional async "waiter" lambda, see custom-resource-handler.py
      logRetention: logs.RetentionDays.ONE_DAY   // default is INFINITE
    });

    const resource = new cdk.CustomResource(this, 'Resource1', { serviceToken: myProvider.serviceToken, properties: props });

    this.response = resource.getAtt('Response').toString();

  }
}

File: java/custom-resource/cdk/src/main/java/software/amazon/awscdk/examples/MyCustomResource.java
package software.amazon.awscdk.examples;

import java.util.Map;
import java.util.UUID;

import software.constructs.Construct;
import software.amazon.awscdk.CustomResource;
import software.amazon.awscdk.Duration;
import software.amazon.awscdk.customresources.*;

import software.amazon.awscdk.services.logs.*;
import software.amazon.awscdk.services.lambda.Runtime;
import software.amazon.awscdk.services.lambda.Code;
import software.amazon.awscdk.services.lambda.SingletonFunction;

public class MyCustomResource extends Construct {
  public String response = "";
  
  public MyCustomResource(final Construct scope, final String id, final Map<String, ? extends Object> props) {
    super(scope, id);

    try {
      final SingletonFunction onEvent = SingletonFunction.Builder.create(this, "Singleton1")
        .uuid(UUID.randomUUID().toString())
        .code(Code.fromAsset("./asset/lambda-1.0.0-jar-with-dependencies.jar"))
        .handler("software.amazon.awscdk.examples.CustomResourceHandler")
        .runtime(Runtime.JAVA_21).memorySize(1024)
        .timeout(Duration.minutes(5))
        .build();

      final SingletonFunction isComplete = SingletonFunction.Builder.create(this, "Singleton2")
        .uuid(UUID.randomUUID().toString())
        .code(Code.fromAsset("./asset/lambda-1.0.0-jar-with-dependencies.jar"))
        .handler("software.amazon.awscdk.examples.CustomResourceIsCompleteHandler")
        .runtime(Runtime.JAVA_21).memorySize(1024)
        .timeout(Duration.minutes(5))
        .build();

      final Provider myProvider = Provider.Builder.create(this, "MyProvider")
        .onEventHandler(onEvent)
        .isCompleteHandler(isComplete)
        .logRetention(RetentionDays.ONE_DAY)
        .build();

      final CustomResource resource = CustomResource.Builder.create(this, "Resource1")
        .serviceToken(myProvider.getServiceToken())
        .properties(props)
        .build();

      response = resource.getAttString("Response");

    } catch (Exception e) {
      e.printStackTrace();
    }
  }
}
